### ISSUE TITLE
Action Handling

### ISSUE THREAD

**Issue by honerlaw** [honerlaw/traph#5]
Most things in the game are actions.

Basically for instance you click on food. You eat the food. Clicking the food is the incoming packet. Eating the food (and healing / removing the food) is the action.

Now we don't want PID to be an issue (specifically for combat). Since combat is based on actions we will use a queue to manage actions.

Basically we handle the incoming packet instantly and create actions based on the incoming packet.

We need to keep track of what players have what actions so we can cancel their actions.

Each entity will have an action manager (this handles their actions). Each action gets register with two things. The action manager (so we can cancel / manage their actions / prioritize them). And the global action queue. The global action queue will be looped through every 600ms.

There can only be one action in the action queue at a time per entity.

Here is an example.

Action One is registered with the action manager / global queue. Action Two is then triggered by the player to be added. Since Action One is not cancellable we completely ignore action two (or send out an error to the user).

Another Example

Action One is registered with the action manager / global queue. Action Two is then triggered by the player. Since Action One is cancellable we remove Action One from the queue and push Action Two onto the end of the queue.

So basically it becomes first come first serve type of deal.
